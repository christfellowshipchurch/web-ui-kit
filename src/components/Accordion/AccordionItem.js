import React, { useState } from 'react';
import classnames from 'classnames'
import PropTypes from 'prop-types';
import {
  accordionItem, accordionItemOpen,
  accordionTitle, accordionIcon, accordionBody, accordionButton,
  contentPadding,
} from '../css/styles.module.css'

const AccordionItem = ({ id, title, isOpen, onClick, children }) => {

  return (
    <div className={classnames(accordionItem, isOpen ? accordionItemOpen : '')} status={isOpen ? 'open' : 'closed'}>
      <button className={accordionButton} onClick={() => onClick(id)}>
        <div className={accordionTitle}>
          {title}
        </div>
        <div className={accordionIcon}>
          <span></span>
          <span></span>
        </div>
      </button>

      <div className={accordionBody}>
        <div className={contentPadding}>
          {children}
        </div>
      </div>
    </div >
  );
}

const propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  isOpen: PropTypes.bool,
  onClick: PropTypes.func
};
const defaultProps = {
  isOpen: false,
  onClick: () => { }
};


AccordionItem.propTypes = propTypes;
AccordionItem.defaultProps = defaultProps;

export default AccordionItem;
