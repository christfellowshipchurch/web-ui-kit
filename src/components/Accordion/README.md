# Accordion

![alt christ-fellowship-flat-ui](/docs/static/Accordion.png)
## About
The accordion is set up with items in 2 columns, each item described with a Title and Content. The Content accepts Html, so you can insert whatever you would like inside of it. Each item has an icon on the right-hand side which is a **"+"** when the item is closed and a **"-"** when the item is opened. There are 4 different configurations for an accordion:

* **Single:** Only 1 item can be opened at a time. Clicking on an open item will close it, and clicking on a closed item will close the opened item, then open itself.

* **Single-Persistent:** Only 1 item can be opened at a time. When an item is open, it will stay open and only close when a closed item it clicked.

* **Multiple:** You are able to open and close as many items in the accordion as you would like.

* **Persistent:** You are able to open as many items in the accordion as you would like. Once an item is opened, it cannot be closed.

## Usage
The Accordion will automatically wrap each child passed in the `AccordionItem` as it renders, so all you need to do is drop in the Html you want to be rendered in each item. Just add the `title` attribute to each container.
```jsx
<Accordion>
    <div title="Item 1">
        <h2>This is one item that I want displayed</h2>
    </div>

    <h2 title="Item 2">Since the Accordion parses each child, there is no need to adhere to a specific structure for the children</h2>

    <ReactComponent title="Item 3">You can even pass in a React Component</ReactComponent>
</Accordion>
```


## Testing
Flat UI Web using jest snapshots for testing.

Running `yarn test` runs a test and outputs the result. When run at the _root_, it will run tests for the entire project. Running at a specific directory will run tests for that directory.

Running `yarn test -u` will update snapshots. **Only do this if you are positive that snapshots need to be updated**.

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate. Commits without tests are not likely to be taken into consideration.

## Authors and acknowledgment
[Caleb Panza](https://github.com/calebpanza), Cristian Garcia

## License
[MIT](https://choosealicense.com/licenses/mit/)