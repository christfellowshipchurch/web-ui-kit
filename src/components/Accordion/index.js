import Wrapper from './Accordion'

export default Wrapper
