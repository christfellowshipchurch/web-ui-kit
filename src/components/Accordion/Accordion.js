import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  includes
} from 'lodash';

import AccordionItem from './AccordionItem';

import {
  accordionContainer
} from '../css/styles.module.css'

const ACCORDION_TYPES = {
  Single: 'single',
  SinglePersistent: 'single-persistent',
  Multiple: 'multiple',
  Persistent: 'persistent',
};

const Accordion = ({ type, children }) => {
  const [sections, setSections] = useState([])

  const toggle = (id) => {
    switch (type) {
      case ACCORDION_TYPES.SinglePersistent:
        setSections([id]);
        break
      case ACCORDION_TYPES.Multiple:
        setSections(
          includes(sections, id)
            ? sections.filter(value => value !== id)
            : [...sections, id]
        )
        break
      case ACCORDION_TYPES.Persistent:
        setSections(
          includes(sections, id)
            ? sections
            : [...sections, id]
        )
        break
      default:
        setSections(
          includes(sections, id)
            ? []
            : [id]
        )
        break
    }
  }

  return (
    <div className={accordionContainer}>
      {children.map((child, i) => {
        let itemId = `section-${i}`

        return (
          <AccordionItem
            id={itemId}
            title={child.props.title}
            isOpen={sections.includes(itemId)}
            onClick={toggle}
            key={i}
          >
            {child}
          </AccordionItem>
        );
      })}
    </div>
  )
}

const propTypes = {
  type: PropTypes.string,
  children: PropTypes.arrayOf(PropTypes.element).isRequired
};
const defaultProps = {
  type: 'single',
};


Accordion.propTypes = propTypes;
Accordion.defaultProps = defaultProps;

export default Accordion;
