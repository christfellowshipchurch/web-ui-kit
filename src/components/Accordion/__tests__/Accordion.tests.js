import React from 'react'
import renderer from 'react-test-renderer'
import {
    configure, shallow, mount, render
} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Accordion from '../../Accordion'

configure({ adapter: new Adapter() });

describe('Accordion', () => {
    it('renders an accordion with items inside', () => {
        const tree = renderer.create(
            <Accordion>
                <div title="Item 1">
                    This is item 1
                </div>
                <div title="Item 2">
                    This is item 2
                </div>
                <div title="Item 3">
                    This is item 3
                </div>
            </Accordion>
        )

        expect(tree).toMatchSnapshot()
    })

    // Type: Single
    it('opens one item at a time with the ability to close an item that is open', () => {
        const wrapper = mount(
            <Accordion type="single">
                <div title="Item 1">
                    This is item 1
                </div>
                <div title="Item 2">
                    This is item 2
                </div>
                <div title="Item 3">
                    This is item 3
                </div>
            </Accordion>
        )
        const firstButton = wrapper.find('button').first()
        const secondButton = wrapper.find('button').at(1)

        // Opens and closes an item
        firstButton.simulate('click')
        wrapper.update()

        let sections = wrapper.find("[status='open']").first()
        expect(sections.length).toEqual(1)

        firstButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']").first()
        expect(sections.length).toEqual(0)


        // Closes an open item after opening a new item
        firstButton.simulate('click')
        wrapper.update()

        secondButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(1)
        expect(sections.find('button').find('div').first().html()).toEqual('<div>Item 2</div>')
    })

    // Type: Single Persisten
    it('opens one item at a time but cannot close an item that is open', () => {
        const wrapper = mount(
            <Accordion type="single-persistent">
                <div title="Item 1">
                    This is item 1
                </div>
                <div title="Item 2">
                    This is item 2
                </div>
                <div title="Item 3">
                    This is item 3
                </div>
            </Accordion>
        )
        const firstButton = wrapper.find('button').first()
        const secondButton = wrapper.find('button').at(1)

        // Opens and closes an item
        firstButton.simulate('click')
        wrapper.update()

        let sections = wrapper.find("[status='open']").first()
        expect(sections.length).toEqual(1)

        firstButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']").first()
        expect(sections.length).toEqual(1)


        // Closes an open item after opening a new item
        firstButton.simulate('click')
        wrapper.update()

        secondButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(1)
        expect(sections.find('button').find('div').first().html()).toEqual('<div>Item 2</div>')
    })

    // Type: Multiple
    it('opens multiple items at a time with the ability to close an item that is open', () => {
        const wrapper = mount(
            <Accordion type="multiple">
                <div title="Item 1">
                    This is item 1
                </div>
                <div title="Item 2">
                    This is item 2
                </div>
                <div title="Item 3">
                    This is item 3
                </div>
            </Accordion>
        )
        const firstButton = wrapper.find('button').first()
        const secondButton = wrapper.find('button').at(1)

        // Opens 2 sections
        firstButton.simulate('click')
        wrapper.update()

        secondButton.simulate('click')
        wrapper.update()

        let sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(2)

        // Closes an open section
        firstButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(1)
    })

    // Type: Persistent
    it('opens multiple items at a time but cannot close an item that is open', () => {
        const wrapper = mount(
            <Accordion type="persistent">
                <div title="Item 1">
                    This is item 1
                </div>
                <div title="Item 2">
                    This is item 2
                </div>
                <div title="Item 3">
                    This is item 3
                </div>
            </Accordion>
        )
        const firstButton = wrapper.find('button').first()
        const secondButton = wrapper.find('button').at(1)

        // Opens 2 sections
        firstButton.simulate('click')
        wrapper.update()

        secondButton.simulate('click')
        wrapper.update()

        let sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(2)

        // Closes an open section
        firstButton.simulate('click')
        wrapper.update()

        sections = wrapper.find("[status='open']")

        expect(sections.length).toEqual(2)
    })
})