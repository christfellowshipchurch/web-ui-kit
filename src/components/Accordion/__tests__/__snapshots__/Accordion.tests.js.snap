// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders an accordion with items inside 1`] = `
<div>
  <div
    className=""
    status="closed"
  >
    <button
      onClick={[Function]}
    >
      <div>
        Item 1
      </div>
      <div>
        <span />
        <span />
      </div>
    </button>
    <div>
      <div>
        <div
          title="Item 1"
        >
          This is item 1
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
    status="closed"
  >
    <button
      onClick={[Function]}
    >
      <div>
        Item 2
      </div>
      <div>
        <span />
        <span />
      </div>
    </button>
    <div>
      <div>
        <div
          title="Item 2"
        >
          This is item 2
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
    status="closed"
  >
    <button
      onClick={[Function]}
    >
      <div>
        Item 3
      </div>
      <div>
        <span />
        <span />
      </div>
    </button>
    <div>
      <div>
        <div
          title="Item 3"
        >
          This is item 3
        </div>
      </div>
    </div>
  </div>
</div>
`;
